saleor/core/utils/metadata_manager.py

```diff
@@ -0,0 +1,72 @@
+from dataclasses import dataclass
+from enum import Enum
+
+from ...graphql.meta.inputs import MetadataInput
+from ..models import ModelWithMetadata
+
+
+class MetadataType(Enum):
+    PUBLIC = "PUBLIC"
+    PRIVATE = "PRIVATE"
+
+
+class MetadataEmptyKeyError(Exception):
+    pass
+
+
+@dataclass
+class MetadataItemCollection:
+    @dataclass
+    class MetadataItem:
+        key: str
+        value: str
+
+        def __init__(self, key: str, value: str):
+            if not key.strip():
+                raise MetadataEmptyKeyError()
+
+            self.key = key
+            self.value = value
```

Comment:
Why you nested `MetadataItem` inside `MetadataItemCollection`? IMO those should be independent classes

Reply:
Agree, it was some part of discussions with @korycins , not sure where we standed. I want to keep MetadataItem class to ensure it's typed. We did have some discussions about keeping 2 classes (which I think is fine)

@korycins WDYT to extract this MetadataItem outside parent class?